### PR TITLE
Solutions are added.

### PR DESCRIPTION
**Pull request by MericCelik**
Same algorithm works with different types of dictionaries. One of them uses Trie data structure in order to make addition and search faster. I could not write many test statements but I am still learning about JUnit, I will add more this week. I hope you like it! :)

### CODE REVIEW

File: MericCelik/assignment3/Dictionary.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        for(String word:words){
+            wordsize = word.length();
+            for(int i=1; i<wordsize+1; i++){
+                prefixset.add(word.substring(0,i));
+            }
+        }
+        return prefixset.toArray(new String[prefixset.size()]);
+    }
+
+    public Dictionary(String[] words){
```
Comment (imorlowska) [inline MericCelik/assignment3/Dictionary.java]: I'd say it would be acceptable to have this constructor as the only public constructor. :) You can remove the other one.

File: MericCelik/assignment3/Dictionary.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import java.util.HashSet;
+
+/**@author AycaMericCelik
+ * */
+
+public class Dictionary {
+    private String[] words;
```
Comment (imorlowska) [inline MericCelik/assignment3/Dictionary.java]: How about using HashSets instead of String[] for both words and prefixes? Then the lookup times are much faster (you could then just do words.contains(word) and prefixes.contains(prefix) instead of writing your own 'contains' method). :) You also don't need to worry about duplicates then.

File: MericCelik/assignment3/Main.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+     * coordinates, grids matrix and the dictionary.
+     * @param dict : The dictionary that contains some words and prefixes.
+     * @param grids :The grids matrix to construct words from its elements.
+     * @return set of possible words(no duplicates)
+     * @throws NullPointerException when the parameters are invalid.
+     * */
+    public static HashSet<String> findWords(String[][] grids, Dictionary dict){
+
+        if(grids == null || dict ==null)
+            throw new NullPointerException("Invalid parameter!");
```
Comment (imorlowska) [inline MericCelik/assignment3/Main.java]: Very nice! How about including which one of them is null in the error message?

File: MericCelik/assignment3/Main.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        int rowcount = grids.length;
+        int columncount = grids[0].length;
+
+        HashSet<String> allresults = new HashSet<>();
+        ArrayList<String> result;
+        for(int r=0;r<rowcount;r++){
+            for(int c=0;c<columncount;c++){
+                int[][] visited = new int[rowcount][columncount];
+                result = buildAllWords(grids,r,c,"",visited,dict, new ArrayList<>());
```
Comment (imorlowska) [inline MericCelik/assignment3/Main.java]: how about allresults.addAll(buildAllWords(...)) or passing allResults as a parameters to buildAllWords method (if you change the last param from ArrayList<String> to HashSet<String>)?

File: MericCelik/assignment3/Main.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+        word += grids[r][c];
+        int rowcount = grids.length;
+        int columncount = grids[0].length;
+
+        if(!dict.isPrefix(word))
+            return result;
+        visited[r][c] = 1;
+
+        if(dict.isWord(word)&&!result.contains(word)){
```
Comment (imorlowska) [inline MericCelik/assignment3/Main.java]: If result was a HashSet<String> you wouldn't need to check for 'contains' here. :)

File: MericCelik/assignment3/Main.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            result.add(word);
+        }
+
+        if(c-1>-1 && visited[r][c-1]==0 && grids[r][c-1]!=null)
+            result = buildAllWords(grids, r, c-1, word, visited, dict, result);  //left = c-1;
```
Comment (imorlowska) [inline MericCelik/assignment3/Main.java]: If you modify `result` in the called method, no need to reassign it here again (that is just calling `buildAllWords(grids, r, c-1, word, visited, dict, result)` without assigning it to `result` should also work if the modifications are done properly).

File: MericCelik/assignment3/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
+import org.junit.Test;
+import org.junit.Assert;
+import java.util.Arrays;
+import java.util.HashSet;
+
+/**@author AycaMericCelik
+ * */
+public class MainTest {
+
+    @Test (expected = NullPointerException.class)
```
Comment (imorlowska) [inline MericCelik/assignment3/MainTest.java]: NPE is usually an unexpected error. How about using IllegalArgumentException or similar?

File: MericCelik/assignment3/Main.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    /** Finds all words that can be created using grids. Calls buildAllWords method for each starting grid.
+     * It creates zero-filled visited matrix for each call of buildAllWords method. It passes current grids
+     * coordinates, grids matrix and the dictionary.
+     * @param dict : The dictionary that contains some words and prefixes.
+     * @param grids :The grids matrix to construct words from its elements.
+     * @return set of possible words(no duplicates)
+     * @throws NullPointerException when the parameters are invalid.
+     * */
+    public static HashSet<String> findWords(String[][] grids, Dictionary dict){
```
Comment (imorlowska) [inline MericCelik/assignment3/Main.java]: Each cell in the grid should only contain one character. Perhaps it would be a good idea to use `char[][]` instead?

File: MericCelik/assignment3/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        TrieBasedDictionary dict = new TrieBasedDictionary(words);
+
+        String[] expectedOutputList = {"car", "card", "cat"};
+        HashSet<String> expectedOutput = new HashSet<>(Arrays.asList(expectedOutputList));
+
+        Assert.assertEquals(expectedOutput,TrieBasedDictionary.findWords(grids, dict));
+    }
+
+
+
```
Comment (imorlowska) [inline MericCelik/assignment3/MainTest.java]: Please remove extra lines.

File: MericCelik/assignment3/Trie.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public Trie(String[] words){
+        this.root = new TrieNode();
+        for(String word:words)
+            this.addWord(word);
+    }
+
+    public void addWord(String word)
+    {
```
Comment (imorlowska) [inline MericCelik/assignment3/Trie.java]: Why the C formatting?

File: MericCelik/assignment3/TrieBasedDictionary.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+     * @param c: current column coordinate/number
+     * @param word : current word it is constructing
+     * @param result : possible words it finds so far
+     * @param visited : binary array to keep track what grids are visited while constructing current word.
+     * @return list of possible words (no duplicates)
+     *
+     * The logic is the same with the Main's but it uses TrieBasedDictionary, so isWord() and isPrefix() method is much
+     * faster.
+     * */
+    public static ArrayList<String> buildAllWords(String[][] grids, int r, int c, String word, int[][] visited,
```
Comment (imorlowska) [inline MericCelik/assignment3/TrieBasedDictionary.java]: Seems like a duplication from the main class. Maybe have TrieBasedDictionary either extend Dictionary or have both implement BasicDictionary interface, so you can avoid this code duplication?

File: MericCelik/assignment3/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
+import org.junit.Test;
+import org.junit.Assert;
+import java.util.Arrays;
+import java.util.HashSet;
+
+/**@author AycaMericCelik
+ * */
+public class MainTest {
```
Comment (imorlowska) [inline MericCelik/assignment3/MainTest.java]: Very nice tests! :) How would you feel about adding some more unit tests for other methods (like isWord and isPrefix in both of your dictionary implementations) or a tests with your own grid and dictionary where you test for different edge-cases (like illegal words with re-using letters, illegal words that would work if jumping over cells was allowed, words on diagonals, reverse diagonals etc.)

File: MericCelik/assignment3/Main.java (hunk 3 above)
Comment (MericCelik) [reply MericCelik/assignment3/Main.java]: I tried to refactor the code to avoid duplicates. Now it takes two other parameter named "triedict" and "isTrieBased". It throws exception in these cases:
- isTrieBased is false, dict is null,
- isTrieBased is true, triedict is null.

In other situations, the algorithm does not use the null dictionary because of the isTrieBased, so it is not a problem. I hope they are clean and healthy ways to do, both refactoring and handling exceptions. :blush:

File: MericCelik/assignment3/Main.java (hunk 4 above)
Comment (MericCelik) [reply MericCelik/assignment3/Main.java]: HashSet is a very precious thing to use, I should have used it more frequently in this assignment! :smile:

File: MericCelik/assignment3/Main.java (hunk 5 above)
Comment (MericCelik) [reply MericCelik/assignment3/Main.java]: Corrected! :+1:

File: MericCelik/assignment3/Main.java (hunk 6 above)
Comment (MericCelik) [reply MericCelik/assignment3/Main.java]: Corrected! Sometimes recursion blows my mind. :smile:

File: MericCelik/assignment3/MainTest.java (hunk 12 above)
Comment (MericCelik) [reply MericCelik/assignment3/MainTest.java]: Thank you so much! I added some more :blush:

File: MericCelik/assignment3/BasicDictionary.java [hunk 13]
```diff
@@ -0,0 +1,5 @@
+
+public interface BasicDictionary {
```
Comment (imorlowska) [inline MericCelik/assignment3/BasicDictionary.java]: Very nice! :)

File: MericCelik/assignment3/MainTest.java (hunk 7 above)
Comment (MericCelik) [reply MericCelik/assignment3/MainTest.java]: Corrected! :+1:

File: MericCelik/assignment3/MainTest.java (hunk 9 above)
Comment (MericCelik) [reply MericCelik/assignment3/MainTest.java]: Done :+1:

File: MericCelik/assignment3/MainTest.java (hunk 12 above)
Comment (imorlowska) [reply MericCelik/assignment3/MainTest.java]: Oh wow, amazing job! So many tests! :D

File: MericCelik/assignment3/Trie.java (hunk 10 above)
Comment (MericCelik) [reply MericCelik/assignment3/Trie.java]: I don't know very much about it, is it the formatting above?
public foo()
{
//code
}

Actually I have always written code like this since couple of months ago. :smile: Why is it called C formatting and is it better to avoid it?

File: MericCelik/assignment3/TrieBasedDictionary.java (hunk 11 above)
Comment (MericCelik) [reply MericCelik/assignment3/TrieBasedDictionary.java]: Merged two functions. :+1: I added one extremely simple interface too! :smile: It helped me to remember the usage of it, thank you!

File: MericCelik/assignment3/Dictionary.java (hunk 2 above)
Comment (MericCelik) [reply MericCelik/assignment3/Dictionary.java]: You are right! Corrected :smile:

File: MericCelik/assignment3/BasicDictionary.java (hunk 13 above)
Comment (MericCelik) [reply MericCelik/assignment3/BasicDictionary.java]: Thank you! :relaxed:
